### ISSUE TITLE
Example Printing

### ISSUE THREAD

**Issue by DRMacIver** [HypothesisWorks/hypothesis-ruby#6]
Currently if a test fails this doesn't actually tell you what example inputs resulted in the failure. This is sub-optimal!

The way this works in hypothesis-python is that we run the whole shrink process etc and then we rerun one final time with `print_examples=True`, which will print all of the top level draw calls (givens in this case). I think that's a sensible approach to emulate.

~~I think this will only make sense to implement after #8 so we can have that final replay. After that it should be "easy", but~~ actually we already replay the final example once, so this is already done.

I need to figure out if there's anything special I should do to get RSpec/etc. to display the output nicely or whether I should just use `puts`.

@samphippen any suggestions on that front?

**Comment by fables-tales** [comment]
@DRMacIver 

Universal protocol in Ruby for failure reporting in a test is to raise an exception, and most test frameworks will directly call `#inspect` on your test failure exception when you raise it, so I'd start by defining `#inspect` on your custom exception classes and see how they work.

**Comment by fables-tales** [comment]
@DRMacIver I just did a quick clone of hypothesis and ran the tests, and got this:

```
  1) some basic hypothesis tests should find that string addition is not commutative
     Failure/Error: expect(x + y).to be == y + x

       expected: == "ba"
            got:    "ab"
```

which is I get only because you have very 'dummy' generators and no example shrinking to speak of right now. I'm happy to contribute the base of a patch that makes the errors format nicely within RSpec once you're a little further along in the implementation if that'll help?

**Comment by DRMacIver** [comment]
> Universal protocol in Ruby for failure reporting in a test is to raise an exception, and most test frameworks will directly call #inspect on your test failure exception when you raise it, so I'd start by defining #inspect on your custom exception classes and see how they work.

Right, but my point is that I want to attach detail information to a test which is not the failure. e.g. in that example I want to show what the values were provided for x and y are.

The plan is to work with arbitrary exceptions rather than requiring any custom exception classes at all. Otherwise it won't work when people are using e.g. matchers as in those tests.

**Comment by fables-tales** [comment]
In that case what you'll probably want to do is implement your own RSpec matcher which formats the failures as you'd expect. RSpec doesn't have a protocol for arbitrary objects to format themselves, yet, unfortunately.

**Comment by DRMacIver** [comment]
> In that case what you'll probably want to do is implement your own RSpec matcher which formats the failures as you'd expect

I really don't think this is a viable solution at all. It's not a question of objects being matched on formatting themselves: The objects that need to be printed don't need to appear anywhere in the matcher at all.

What does rpsec do with things printed to stdout during a test? Does it do any sort of output capturing? Because right now "Just call puts" is sounding like a pretty attractive solution.

**Comment by DRMacIver** [comment]
So to be clear, what I want is to have a test something like this:

```ruby
it "does stuff" do
   hypothesis do
      x = given integers
      y = given integers
      a, b = some_complicated_computations(x, y)
      expect(a).to eq(b)
   end
end
```

And when this fails I want to a) Show the normal rspec matcher error and b) Print the values of x and y as part of the test failure.

**Comment by DRMacIver** [comment]
> I really don't think this is a viable solution at all

Mmm. Those words are delicious with salt.
